### PR TITLE
NodeJS Autocomplete search bar

### PR DESCRIPTION
**Pull request by proudlygeek**
# NodeJS Api doc autocomplete

This is a proposal for adding an Autocomplete searchbox on NodeJS documentation page.
- Added autocomplete.js
- CSS classes for different types
- Link to specific pages
- Favicon
- Feedback integration from the team :smile:
- New hierarchy strategy
- Modified JSON parser to parse properties
- Removed parameters from index
- Removed code blocks from description
- Breadcrumbs
- Breadcrumb can now be clicked
- Fixed globals bug

### CODE REVIEW

File: doc/api_assets/api-realtime-search.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    templates: {
+      empty: function(data) {
+        return '<div class="no-results">We didn\'t find any results for "' + data.query + '". Sorry!</div>';
+      },
+      suggestion: function(suggestion) {
+        var stability = (suggestion.isStable) ? 'stable' : 'deprecated';
+        var showDescription = (suggestion._highlightResult.textValue.matchedWords.length) ? false : true;
+        var breadcrumb = suggestion.breadcrumb || '';
+
+        if (showDescription) {
```
Comment (redox) [inline doc/api_assets/api-realtime-search.js]: I think we could factorize those 2 blocks and just add the `description` tags+content if needed.

File: tools/doc/json.js [hunk 2]
```diff
@@ -215,7 +215,8 @@ function processList(section) {
   // shove the name in there for properties, since they are always
   // just going to be the value etc.
   if (section.type === 'property' && values[0]) {
-    values[0].textRaw = '`' + section.name + '` ' + values[0].textRaw;
+    //values[0].textRaw = '`' + section.name + '` ' + values[0].textRaw;
```
Comment (redox) [inline tools/doc/json.js:218]: We'll need to explain that change.

File: tools/doc/syncIndexes.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+function readAllJson(err, data) {
+  if (err) {
+    throw err;
+  }
+
+  var object  = JSON.parse(data), 
+      indexes = chunk(fold(object, [], 0, markdownName(object.source), []), 5000, null);
+
+  if (process.argv.length > 1 && process.argv[2] == '--dummy') {
```
Comment (redox) [inline tools/doc/syncIndexes.js]: I would rename `--dummy` to `--debug` and maybe add a usage if `process.argv.length > 2` or `--help` ?

**Comment by redox** [comment]
I think we'll also need to squash the commit messages together; no need to have the whole detail. But very good job @proudlygeek !

**Comment by proudlygeek** [comment]
Thanks @redox :) good to get some feedback on this!

File: tools/doc/json.js (hunk 2 above)
Comment (proudlygeek) [inline tools/doc/json.js:218]: This is the trickiest part, the JSON parser is definitely old and not documented. I think we can add a property without editing `textRaw`, but I have to try it yet. I'll rather not modify or explain this change :smile:

File: doc/api_assets/api-realtime-search.js (hunk 1 above)
Comment (proudlygeek) [inline doc/api_assets/api-realtime-search.js]: Right on, that's a little bit rushed! :+1:

File: tools/doc/syncIndexes.js (hunk 3 above)
Comment (proudlygeek) [inline tools/doc/syncIndexes.js]: Makes sense!

**Pull request closed without merging**